File: ThreatModelDfdService/Services/Impl/ProjectService.cs
using Microsoft.EntityFrameworkCore;
using ThreatModelDfdService.Data.DTO;
using ThreatModelDfdService.Model.Context;
using ThreatModelDfdService.Model.Entity;
using ThreatModelDfdService.Repositories;

namespace ThreatModelDfdService.Services.Impl;

public class ProjectService(
    MSSQLContext _context,
    IRepository<Project> projectRepository,
    DfdService dfdService
    )
{
    public ProjectResponseDTO CreateProject(CreateProjectDTO dto)
    {
        Project project = CreateNewProjectEntity(dto);
        
        Dfd contextDiagram = dfdService.Create(0, project.Id);
        project.ContextDiagramId = contextDiagram.Id;
        projectRepository.Update(project);

        return new ProjectResponseDTO(
            project.Id,
            project.Title,
            project.Description,
            project.ContextDiagramId,
            project.CreatedAt
        );
    }

    public ProjectResponseDTO UpdateProject(long id, UpdateProjectDTO dto)
    {
        Project project = FindById(id);
        project.Title = dto.Name;
        project.Description = dto.Description;
        projectRepository.Update(project);
        return new ProjectResponseDTO(
            project.Id,
            project.Title,
            project.Description,
            project.ContextDiagramId,
            project.CreatedAt
        );
    }

    public async Task<PagedProjectResponseDTO> GetPagedProjectsAsync(int page, int size)
    {
        var totalItems = await _context.Projects.CountAsync();
        var totalPages = (int)Math.Ceiling(totalItems / (double)size);

        var pagedList = await _context.Projects
            .AsNoTracking()
            .OrderByDescending(p => p.CreatedAt)
            .Skip((page - 1) * size)
            .Take(size)
            .ToListAsync();

        var projects = pagedList.Select(p => new ProjectResponseDTO(
            p.Id,
            p.Title,
            p.Description,
            p.ContextDiagramId,
            p.CreatedAt
        )).ToList();

        return new PagedProjectResponseDTO(page, totalPages, projects);
    }

    public async Task DeleteProjectAsync(long id)
    {
        Project project = FindById(id);
        projectRepository.Delete(project);
    }

    private Project CreateNewProjectEntity(CreateProjectDTO dto)
    {
        Project project = new Project
        {
            Title = dto.Name,
            Description = dto.Description,
            CreatedAt = DateTime.UtcNow
        };
        return projectRepository.Create(project);
    }

    public ProjectResponseDTO GetProjectById(long id)
    {
        Project project = FindById(id);
        return new ProjectResponseDTO(
            project.Id,
            project.Title,
            project.Description,
            project.ContextDiagramId,
            project.CreatedAt
        );
    }

    public Project FindById(long id)
    {
        Project? project = projectRepository.FindById(id);
        if (project == null) throw new KeyNotFoundException(
            "Project not found with the provided ID: " + id);
        return project;
    }
}

File: dfd-service/src/main/java/com/tcc/dfd_service/service/ProjectService.java
package com.tcc.dfd_service.service;

import com.tcc.dfd_service.dto.*;
import com.tcc.dfd_service.entity.Dfd;
import com.tcc.dfd_service.entity.Project;
import com.tcc.dfd_service.repository.ProjectRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.time.OffsetDateTime;
import java.time.ZoneOffset;

@Service
@RequiredArgsConstructor
public class ProjectService {

    private final ProjectRepository projectRepository;
    private final DfdService dfdService;

    public ProjectResponseDTO createProject(CreateProjectDTO dto) {

        Project project = createNewProjectEntity(dto);

        Dfd contextDiagram = dfdService.create(0, project.getId(), null);

        project.setContextDiagramId(contextDiagram.getId());

        project = projectRepository.save(project);
        return toResponse(project);
    }

    public ProjectResponseDTO updateProject(Long id, UpdateProjectDTO dto) {

        Project project = findById(id);

        project.setTitle(dto.name());
        project.setDescription(dto.description());

        project = projectRepository.save(project);
        return toResponse(project);
    }

    public PagedProjectResponseDTO getPagedProjects(Integer page, Integer size) {
        if(page == null) page = 0;
        if(size == null) size = 15;
        Pageable pageable = PageRequest.of(page,size, Sort.by("createdAt").descending());

        Page<Project> pagedProjects =
                projectRepository.findAll(pageable);

        return new PagedProjectResponseDTO(
                page,
                pagedProjects.getTotalPages(),
                pagedProjects.getContent()
                        .stream()
                        .map(this::toResponse)
                        .toList()
        );
    }

    public void deleteProject(Long id) {
        Project project = findById(id);
        projectRepository.delete(project);
    }

    public ProjectResponseDTO getProjectById(Long id) {
        return toResponse(findById(id));
    }

    public Project findById(Long id) {
        return projectRepository.findById(id)
                .orElseThrow(() ->
                        new IllegalArgumentException(
                                "Project not found with the provided ID: " + id
                        ));
    }

    private Project createNewProjectEntity(CreateProjectDTO dto) {

        Project project = new Project();

        project.setTitle(dto.name());
        project.setDescription(dto.description());
        project.setCreatedAt(
                OffsetDateTime.now(ZoneOffset.UTC)
        );

        return projectRepository.save(project);
    }

    private ProjectResponseDTO toResponse(Project project) {

        return new ProjectResponseDTO(
                project.getId(),
                project.getTitle(),
                project.getDescription(),
                project.getContextDiagramId(),
                project.getCreatedAt()
        );
    }
}